clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -497,14 +517,22 @@ public boolean hasNext() {
         @Override
         public Record next() {
             if (readRecords >= numRecords)
-                throw new NoSuchElementException();
+                throw new InvalidRecordException("Garbled data resulting in attempt of reading past the end of batch");
 
             readRecords++;
-            return readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            Record rec = readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            if (readRecords == numRecords) {
+                // validate the batch
```

Comment:
nit: this comment doesn't seem too useful. Maybe emphasize that this check ensures that the declared size of the batch is correct?